FOE support

jokva opened the pull request:
Field efficiency: (OIP(initial) - OIP(now)) / OIP(initial). The initial
OIP is cached, and FOIP is reused for OIP(now). Adds the
operator-(double,quantity) to make the formula obvious from the FOE
function.

@atgeirr @babrodtk I'd appreciate some help on the simulator side of this. Somewhere around https://github.com/OPM/opm-simulators/blob/master/opm/autodiff/SimulatorBase_impl.hpp#L190 you'd need to call `setInitial` with OIP passed.

jokva committed:
FOE support

Field efficiency: (OIP(initial) - OIP(now)) / OIP(initial). The initial
OIP is cached, and FOIP is reused for OIP(now). Adds the
operator-(double,quantity) to make the formula obvious from the FOE
function.

pgdr reviewed (approved):
This looks good and ready to be merged.  But perhaps we should let Atgeirr and André chime in before merging.

atgeirr commented:
I'll ask @babrodtk to look at this, I need to concentrate on the Frankenstein merge.

jokva commented:
It won't really do any new output (save writing zeros) until `setInitial` is called, so it's sort-of harmless to merge, since simulator support can be added independently.

pgdr commented:
I agree it's safe to merge.

However, this is an "API change (or addendum)" and therefore; perhaps the users of the API would have opinions.

jokva commented:
Absolutely.

jokva commented:
I've updated this slightly and removed the `setInitial` method. Instead, you need  to pass "OIP" to `writeInitAndEgrid`, which has now been renamed `writeInitial`. There might be more properties later that requires some knowledge of the initial state, and it's likely that they too will need information through `writeInitial`.

This patch will fail on travis now - when the simulator is ready to adopt this, please merge (even if you're not actually passing OIP). Norne and friends don't use the `FOE` keyword, so it won't break any regression test through keyword support alone.

joakim-hove commented on opm/output/eclipse/Summary.cpp:
```diff
@@ -801,6 +819,13 @@ void Summary::add_timestep( int report_step,
     this->prev_time_elapsed = secs_elapsed;
 }
 
+void Summary::set_initial( const data::Solution& sol ) {
```
I don't know if we will need more of these initial properties; but at least for the second case I think we should use a map-like structure to internalize these initial values.

jokva replied:
That's my plan if/when that shows up, as this solution obviously doesn't scale (but is delightfully simple).

joakim-hove replied:
OK - I am also a proponent of jit bridge burning.

joakim-hove commented:
Will you make a `s/writeEgridandInit/writeInitial/` PR in opm-simulators?

jokva commented:
Yeah, I can do that.
